### DISCUSSION TITLE
Preserving previous location while authenticating

### DISCUSSION THREAD

**Discussion by jabidof** [TanStack/router#125] [Q&A] [answered]
I'm trying to preserve the **previous location** so I can send the user there **after** they authenticate.
The use case is described in below sequence

1. Non-authorized user enter any URL pointing to a particular route:
`.../dashboard/topics/123?readMode=raw`    <- save this!
2. User is re-routed to `.../login`
3. User logged in successfully and is navigated to "saved URL"

What is the _best practice_ using `react-location`?

**Comment by tannerlinsley** [comment, marked as answer]
You can capture the location from `useLocation` or `useRouter` into something persistent like localStorage or even a `?redirect=url` search param, then either declaratively `<Navigate>` or imperatively `useNavigate` to a new route. When things are good, do the same back to the original location you captured.

**Reply by caturhidayat** [reply]
Dear @tannerlinsley ,

can you give me some examples(script) for this secenario?

**Reply by V1RE** [reply]
@caturhidayat There's an example in the docs here: [Overview | Router Docs](https://tanstack.com/router/v1/docs/guide/authenticated-routes#redirecting)

**Reply by caturhidayat** [reply]
@V1RE I don't get in about : 
`router.history.push(search.redirect)` // How to use it

can you give me some examples?

**Comment by jabidof** [comment]
Thanks for the info.
I just implemented the mechanism using `?redirect=url` approach as described in #92. All good!

Just my two cents regarding the doc (which is already close to perfection (Y)). The `useLocation` is nowhere mentioned unless I'm wrong?

**Comment by eastack** [comment]
I have a similar issue which is after submitting the form or clicking the cancel button, how can I jump back to the previous page, or a few pages back.

**Comment by eastack** [comment]
I tried to use useNavigate to implement it but it seems I have to provide a specific `to` parameter.
